ledger/participant-integration-api/src/main/scala/platform/indexer/ha/HaCoordinator.scala

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer.ha
+
+import java.sql.Connection
+
+import akka.actor.Scheduler
+import akka.stream.KillSwitch
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.backend.DBLockStorageBackend.{Lock, LockId, LockMode}
+import com.daml.platform.store.backend.DBLockStorageBackend
+import javax.sql.DataSource
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+/** A handle of a running program
+  * @param completed will complete as the program completes
+  *                  - if no KillSwitch used,
+  *                    - it will complete successfully as program successfully ends
+  *                    - it will complete with a failure as program failed
+  *                  - if KillSwitch aborted, this completes with the same Throwable
+  *                  - if KillSwitch shut down, this completes successfully
+  *                  As this completes, the program finished it's execution, also all resources released as well.
+  * @param killSwitch to signal abortion and shutdown
+  */
+case class Handle(completed: Future[Unit], killSwitch: KillSwitch)
+
+/** This functionality sign off a worker Connection, anc clears it for further usage.
+  * This only need to be done once at the beginning of the Connection life-cycle
+  * On any error an exception will be thrown
+  */
+trait SignConnection {
+  def sign(connection: Connection): Unit
+}
+
+/** To add High Availability related features to a program
+  */
+trait HaCoordinator {
+
+  /** Execute block in High Availability mode.
+    * Wraps around the Handle of the block.
+    *
+    * @param block HaCoordinator provides a SignConnection which need to be used for all database connections to do work in the block
+    *              Future[Handle] embodies asynchronous initialisation of the block
+    *              (e.g. not the actual work. That asynchronous execution completes with the completed Future of the Handle)
+    * @return the new Handle, which is available immediately to observe and interact with the complete program here
+    */
+  def protectedBlock(block: SignConnection => Future[Handle]): Handle
+}
+
+case class HaConfig(
+    mainLockAquireRetryMillis: Long = 500,
+    workerLockAquireRetryMillis: Long = 500,
+    workerLockAquireMaxRetry: Long = 1000,
+    mainLockCheckerPeriodMillis: Long = 1000,
```

Comment:
For one: this is not entirely true:
* Until fail-over is done, quite some other work is needed, so the total fail over time is probably much longer
* Tuning this guy only helps in a corner scenario, where on node1 just the main connection is lost, and node1 worker connections still grab on the shared worker lock-s. In this situation node2 can already grab the main lock, and is already in the loop for getting the exclusive worker lock, so this case also workerLockAquireRetryMillis would need adapting, to achive the desirable effect.
* This corner scenario above is quite unlikely, much more likely is that the connection is severed entirely, and all connections (therefore locks) released promptly. In this case it is indifferent what happens on node1, interesting would be mainly the mainLockAquireRetryMillis, which controls on node2 how fast it can acquire the main lock, and start the sequence from there.

Also for testing purposes I would prefer to keep this not too frequent, so we are not littering the logs aggressively.
Would you be okay to tune this (and preferably other durations here) for HA deployed scenarios with config overrides?

Reply:
Thanks for the detailed explanation. Let's tune this via config and leave the 1s default.